core/index/index_writer.cpp

```diff
@@ -64,67 +68,53 @@ const FeatureInfoProvider kDefaultFeatureInfo = [](irs::type_info::type_id) {
                    FeatureWriterFactory{}};
 };
 
-struct FlushSegmentContext {
-  // starting doc_id to consider in 'segment.meta' (inclusive)
-  const doc_id_t doc_id_begin_;
-  // ending doc_id to consider in 'segment.meta' (exclusive)
-  const doc_id_t doc_id_end_;
-  // copy so that it can be moved into 'index_writer::pending_state_'
-  IndexSegment segment_;
-  // doc_ids masked in SegmentMeta
-  document_mask docs_mask_;
-  doc_map old2new_;
-  doc_map new2old_;
-  // modification contexts referenced by 'update_contexts_'
-  std::span<const IndexWriter::ModificationContext> modification_contexts_;
-  // update contexts for documents in SegmentMeta
-  std::span<const segment_writer::update_context> update_contexts_;
-  std::shared_ptr<const SegmentReaderImpl> reader_;
-
-  FlushSegmentContext(
-    std::shared_ptr<const SegmentReaderImpl>&& reader, IndexSegment&& segment,
-    document_mask&& docs_mask, doc_id_t doc_id_begin, doc_id_t doc_id_end,
-    std::span<const segment_writer::update_context> update_contexts,
-    std::span<const IndexWriter::ModificationContext> modification_contexts,
-    doc_map&& old2new)
-    : doc_id_begin_{doc_id_begin},
-      doc_id_end_{doc_id_end},
-      segment_{std::move(segment)},
-      docs_mask_{std::move(docs_mask)},
-      old2new_{std::move(old2new)},
-      modification_contexts_{modification_contexts},
-      update_contexts_{update_contexts},
-      reader_{std::move(reader)} {
-    IRS_ASSERT(reader_ != nullptr);
-    IRS_ASSERT(update_contexts.size() == segment_.meta.docs_count);
-    IRS_ASSERT(segment_.meta.docs_count < doc_limits::eof());
-    const auto first_doc_id = doc_limits::min();
-    const auto last_doc_id =
-      first_doc_id + static_cast<doc_id_t>(segment_.meta.docs_count);
-    IRS_ASSERT(doc_id_begin_ < doc_id_end_);
-    IRS_ASSERT(first_doc_id <= doc_id_begin_);
-    IRS_ASSERT(doc_id_end_ <= last_doc_id);
-
-    auto update_docs_mask = [&](auto&& visitor) {
-      for (doc_id_t doc_id = first_doc_id; doc_id < doc_id_begin_; ++doc_id) {
-        docs_mask.emplace(visitor(doc_id));
+struct FlushedSegmentContext {
+  FlushedSegmentContext(std::shared_ptr<const SegmentReaderImpl>&& reader,
+                        IndexWriter::SegmentContext& segment,
+                        IndexWriter::FlushedSegment& flushed)
+    : reader{std::move(reader)}, segment{segment}, flushed_{flushed} {
+    IRS_ASSERT(reader != nullptr);
+    if (!flushed_.old2new.empty() && flushed_.old2new.empty()) {
+      flushed_.new2old.resize(flushed_.old2new.size());
+      for (doc_id_t old_id = 0; const auto new_id : flushed_.old2new) {
+        flushed_.new2old[new_id] = old_id++;
       }
-      for (doc_id_t doc_id = doc_id_end_; doc_id < last_doc_id; ++doc_id) {
-        docs_mask.emplace(visitor(doc_id));
-      }
-    };
+    }
+    const auto begin = flushed.GetDocsBegin();
```

Comment:
segment_writer
// rollback bitset 

index_writer
commit
// end rollback bitset
// remove c1 new_id, tick old_id

c1
// tick removes from bitset
// translate to hash_set
// remove bitset
// remove map new2old
// c1 -> flat_hash_set

c2
// remove c2 new_id, tick old_id hash_set
// c2 -> flat_hash_set